// sanity/lib/structure.esm.js
import { C, h, O, a, x, j, D, m, o, g, c, d, q, Q, G, R, J, I, L, z, M, H, P, b, S, V, N, i, t, y, k, l, K, B, E, n, p, v, A, F, T, w, r, s, U, f, e, u, W } from "./_chunks-es/StructureToolProvider.js";
import "react/jsx-runtime";
import "@sanity/icons";
import "@sanity/ui";
import "react";
import "./_chunks-es/getJsonStream.js";
import "sanity";
import "@sanity/telemetry/react";
import "lodash/template.js";
import "lodash/isEqual.js";
import "@sanity/uuid";
import "sanity/router";
import "react-rx";
import "rxjs";
import "@sanity/types";
import "@sanity/util/paths";
import "lodash/omit.js";
import "lodash/isNumber.js";
import "lodash/isString.js";
import "lodash/throttle.js";
import "scroll-into-view-if-needed";
import "lodash/findIndex.js";
import "@sanity/diff-match-patch";
import "@portabletext/react";
export {
  C as CommentsIntentProvider,
  h as ComponentBuilder,
  O as ComponentViewBuilder,
  a as ConfirmDeleteDialog,
  x as DEFAULT_INTENT_HANDLER,
  j as DocumentBuilder,
  D as DocumentInspectorHeader,
  m as DocumentListBuilder,
  o as DocumentListItemBuilder,
  g as DocumentListPane,
  c as DocumentPane,
  d as DocumentPaneProvider,
  q as DocumentTypeListBuilder,
  Q as FormViewBuilder,
  G as GenericListBuilder,
  R as GenericViewBuilder,
  J as HELP_URL,
  I as InitialValueTemplateItemBuilder,
  L as ListBuilder,
  z as ListItemBuilder,
  M as MenuItemBuilder,
  H as MenuItemGroupBuilder,
  P as PaneLayout,
  b as PaneRouterContext,
  S as SerializeError,
  V as StructureToolProvider,
  N as component,
  i as createStructureBuilder,
  t as defaultInitialValueTemplateItems,
  y as defaultIntentChecker,
  k as documentFromEditor,
  l as documentFromEditorWithInitialValue,
  K as form,
  B as getOrderingMenuItem,
  E as getOrderingMenuItemsForSchemaType,
  n as getTypeNamesFromFilter,
  p as isDocumentListItem,
  v as maybeSerializeInitialValueTemplateItem,
  A as maybeSerializeMenuItem,
  F as maybeSerializeMenuItemGroup,
  T as maybeSerializeView,
  w as menuItemsFromInitialValueTemplateItems,
  r as shallowIntentChecker,
  s as structureLocaleNamespace,
  U as structureTool,
  f as useDocumentPane,
  e as useDocumentTitle,
  u as usePaneRouter,
  W as useStructureTool
};
